File: leetcode/121/MaxProfit.cpp
// #include <bits/stdc++.h>
#include <iostream>
#include <vector>
using namespace std;

// class Solution
// {
// public:
//     int maxProfit(vector<int> &prices)
//     {
//         if (prices.size() == 0)
//         {
//             return 0;
//         }
//         int min_price = prices[0];
//         int maxprofit = 0;
//         for (int i = 0; i < prices.size(); i++)
//         {
//             min_price = min(min_price, prices[i]);
//             maxprofit = max(maxprofit, prices[i] - min_price);
//         }

//         return maxprofit;
//     }
// };

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int min_price = INT_MAX;
        int max_profit = 0;
        
        for(int i = 0; i < prices.size(); i++){
            min_price = min(min_price, prices[i]);
            max_profit = max(max_profit, prices[i]-min_price);
        }
        
        return max_profit;
    }
};

int main()
{
    Solution sl;
    vector<int> vect;

    vect.push_back(7);
    vect.push_back(1);
    vect.push_back(5);
    vect.push_back(3);
    vect.push_back(6);
    vect.push_back(4);

    cout << sl.maxProfit(vect);
    return 0;
}
//gcc -o MaxProfitCpp MaxProfit.cpp -lstdc++

File: leetcode/121/MaxProfit.py
from re import S


class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        min_price = prices[0]
        profit = 0

        for price in prices:
            # min_price = min(min_price, price)
            # profit = max(profit, price-min_price)

            if min_price > price:
                min_price = price
            
            if profit < price-min_price:
                profit = price-min_price

        return profit

print(Solution().maxProfit([ 7,1,5,3,6,4]))